client/landing/gutenboarding/onboarding-block/create-site/create-and-redirect.tsx

```diff
@@ -0,0 +1,50 @@
+/**
+ * External dependencies
+ */
+import React, { useEffect, useState } from 'react';
+import { Redirect } from 'react-router-dom';
+import { useDispatch, useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { STORE_KEY as ONBOARD_STORE } from '../../stores/onboard';
+import { USER_STORE } from '../../stores/user';
+import { useFreeDomainSuggestion } from '../../hooks/use-free-domain-suggestion';
+import { Step, usePath } from '../../path';
+
+type Status = 'INIT' | 'SUCCESS' | 'ERROR' | 'MISSING_SITE_DATA';
+
+const CreateAndRedirect = () => {
+	const { siteTitle, siteVertical } = useSelect( select => select( ONBOARD_STORE ).getState() );
+	const [ status, setStatus ] = useState< Status >( siteVertical ? 'INIT' : 'MISSING_SITE_DATA' );
+
+	const currentUser = useSelect( select => select( USER_STORE ).getCurrentUser() );
+	const makePath = usePath();
+	const { createSite } = useDispatch( ONBOARD_STORE );
+
+	const freeDomainSuggestion = useFreeDomainSuggestion();
+
+	useEffect( () => {
+		// If there's no site title don't wait for a free domain suggestion, there won't be one
+		if ( currentUser && ( ! siteTitle || freeDomainSuggestion ) ) {
+			createSite( currentUser.username, freeDomainSuggestion ).then( success => {
```

Comment:
Also I get an error/warning in the console, I think it is related to this line:
```
 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    in CreateAndRedirect (created by CreateSite)
    in div (created by CreateSite)
    in CreateSite (created by OnboardingEdit)
```

Reply:
Hmmm that callback getting called after we dismount is a problem. In fact I think it might because there's another part of the app listening to store changes and redirecting like you suggested: https://github.com/Automattic/wp-calypso/blob/179848f706f9b34aad297039431522d42e43c549/client/landing/gutenboarding/onboarding-block/edit.tsx#L32

Ideally I could just rely on the redirect in `<OnboardingEdit>`, but it pushes onto the history, but in the frankenflow I really don't want to push. I'll try figure out a way I can reuse this behaviour without pushing.

Reply:
In any flow, we don't want the back button to create another site so maybe the OnboardingEdit should be changed to not push? and then it sounds like we could use that for the create-and-redirect case too. ( I don't fully understand what's happening with the redirects TBH 😅)

Reply:
> I don't fully understand what's happening with the redirects TBH 😅

It's a rabbit warren lol

I've got a commit that uses the redirect in `edit.tsx` to replace `/create-site?new` with `/create-site`.

I think we can address the issue of clicking back from the block editor in another PR. We navigate to the block editor using `window.location.href = ...` and I think it'll work better with `window.location.replace(...)`. We want the blocked editor to _replace_ the site creation animation; if the click back it seems they should go right back to the last real page they saw e.g. design picking.

Reply:
#40359 should stop the back button going back to the animation

Reply:
Ahh nice one in #40359.
Were you able to reproduce/ track down what was throwing that error?:
`Warning: Can't perform a React state update on an unmounted component.`

Reply:
No but this is what I think was happening:
- Navigate to `/create-site?new`
- `<CreateAndRedirect>` renders and starts a request to create a site
- `<Edit>` sees that a site has started to be created and re-renders, redircting to `/create-site`
- `<CreateAndRedirect>` unmounts
- Time passes
- Eventually the call to create a site succeeds, the promise is resolved, so control flow is passed to the promise handler inside the now unmounted `<CreateAndRedirect>`
- Promise handler calls `setStatus` on an unmounted component; error!